R/art.con.internal.R

```diff
@@ -0,0 +1,320 @@
+### grouping.variables: list of grouping variables (of type name) (e.g. list(1|d))
+### error.variables: list of error variables (of type name) (e.g. list(Error(g)))
+#' @importFrom plyr laply
+parse.art.model.formula = function(m.f, f.parsed){
+  
+  # get model formula terms
+  m.f.terms = terms(m.f)
+  # char vector of names of rhs terms and their interactions
+  # e.g. Response ~ a*b + (1|d) + Error(g) -> c("a", "b", "1 | d", "Error(g)", "a:b")) not necessarily in this order
+  m.f.term.labels = attr(m.f.terms, "term.labels")
+  
+  # check if interaction term from f is in model formula
+  f.interaction.term.label = f.parsed$interaction.term.label # f.parsed always has exactly one interaction
+  is.interaction.term = grepl(f.interaction.term.label, m.f.term.labels)
+  
+  # if interaction term from f is not in model formula, error
+  if(!any(is.interaction.term)){
+    stop("Term or formula passed to art.con or artlm.con must contain a single interaction term and no other terms, and the interaction term must be in art model formula.")
+  }
+  
+  # response ~ a*b*c*d + (1|d) + Error(g) -> list(response, a, b, c, d, 1|d, Error(g))
+  m.f.variables.all = as.list(attr(m.f.terms, "variables"))[c(-1)]
+  m.f.response = m.f.variables.all[[1]]
+  m.f.variables = m.f.variables.all[c(-1)]
+  
+  #determine which variables on the rhs are grouping variables, error variables, or fixed variables
+  is.grouping.variable = laply(m.f.variables, function(term) as.list(term)[[1]] == quote(`|`))
+  is.error.variable = laply(m.f.variables, function(term) as.list(term)[[1]] == quote(`Error`))
+  #all other variables that aren't grouping or error variables must be fixed variables
+  is.fixed.variable = !(is.grouping.variable | is.error.variable)
+  
+  m.f.grouping = m.f.variables[is.grouping.variable]
+  m.f.error = m.f.variables[is.error.variable]
+  m.f.fixed = m.f.variables[is.fixed.variable] # remove response
+  
+  list(
+    response = m.f.response,
+    fixed.variables = m.f.fixed,
+    grouping.variables = m.f.grouping,
+    error.variables = m.f.error
+  )
+}
+
+### creates new data frame which is a copy of df except
+###   adds a new column by concatenated columns of df whose names are the interaction variables in f
+###   removes columns whose names are the interaction variables in f
+### m.formula is the original formula used to create the ART model
+### df is the data frame used to creat the ART model
+### formula is the contrast formula
+#' @importFrom tidyr unite_
+generate.art.concatenated.df = function(m.f.parsed, df, f.parsed){
+  
+  # concatenate columns of data frame whose columns names are the variables in f.parsed.interaction.variables
+  # e.g. abc
+  f.concatenated.variable = f.parsed$concat.interaction.variable
+  # e.g. list(a, b, c)
+  f.interaction.variables = f.parsed$interaction.variables
+  art.con.df = unite_(df, f.concatenated.variable, f.interaction.variables, sep = ",", remove = TRUE) # concat columns and remove originals
```

Comment:
`tidyr::unite_` is deprecated and should not be used in new code. Also, if we can replace it we can skip an import :).

This is slightly more annoying to do in base R but can be done by `paste'ing columns together. Here's an example:

```r
df = data.frame(x = paste0("x", 1:5), y = paste0("y", 1:5), z = paste0("z", 1:5))
df
```
```
   x  y  z
1 x1 y1 z1
2 x2 y2 z2
3 x3 y3 z3
4 x4 y4 z4
5 x5 y5 z5
```
```r
variables_to_unite = c("x", "y", "z")
new_col_name = "xyz"

# paste values together into a new column
df[[new_col_name]] = do.call(paste, c(unname(df[,variables_to_unite]), sep = ","))
df
```
```
   x  y  z      xyz
1 x1 y1 z1 x1,y1,z1
2 x2 y2 z2 x2,y2,z2
3 x3 y3 z3 x3,y3,z3
4 x4 y4 z4 x4,y4,z4
5 x5 y5 z5 x5,y5,z5
```
```r
# remove old columns
df[,variables_to_unite] = NULL
df
```
```
       xyz
1 x1,y1,z1
2 x2,y2,z2
3 x3,y3,z3
4 x4,y4,z4
5 x5,y5,z5
```